README.md
# Modified Python Ant Downloader
Modifications were made to make the tool easier for more quickly understanding interactions with ANT devices. These changes don't really help the standard use case of Python Ant Downloader (but will be helpful if you are trying to build upon this great platform to do other intersting things with ANT+ Networks).
--- Modifications to add various Data Pages including RequestPage, HeartRatePages ---
--- Modifications to add much more human readable logging that does lookups on message ids, codes etc and dissects 'data' payloads and applies a page unpacker on them ---

e.g.
[Thread-2]	2013-08-09 16:55:25,410	DEBUG	Processing reply. RECV_BROADCAST_DATA(channel_number=0, data=0x4301000301000200 -> Beacon{'auth_type': 3, 'pairing_enabled': 0, 'data_available': 0, 'device_state': 0, 'period': 1, 'descriptor': 131073, 'upload_enabled': 0, 'data_page_id': 67, 'data': '', 'status_2': 0, 'status_1': 1})
[Thread-1]	2013-08-09 16:55:09,444	DEBUG	Processing reply. CHANNEL_EVENT(channel_number=0, msg_id=MESG_NETWORK_KEY_ID[70], msg_code=RESPONSE_NO_ERROR[0])


# Python Ant Downloader

Tools for extracting data from Garmin wireless (ANT) GPS devices. The project goal is to support downloading data from GPS unit and upload to Garmin Connect. It doesn't support support workout, profile, or activitly uploads like the Windows "Garmin ANT Agent."

This software implements the [Garmin Device Interface Spec](http://www8.garmin.com/support/commProtocol.html) over an [ANT-FS](http://www.thisisant.com) transport. In theory it should work with any device implementing this stack. Early wireless Garmin devices should be supported, but newer hardware uses a different protocol. See "Supported Devices" below.

The software can be run as either a daemon or on-demand. In deamon mode it automatically saves TCX files to a configured directory whenever a paired devices is within range and has new data. In on-demand mode the program just downloads once and terminates. The software also supports automatic upload to Garmin Connect.

## Getting Help

Issues Tracker: https://github.com/braiden/python-ant-downloader/issues

## Supported Devices

So far this software software has been reported to work with:

  * 405
  * 405CX
  * 410

## Unsupported Devices

  * 610 
  * 310 
  * FR60
  * 910XT

These devices (and probably anything newer) appear to implement ANT-FS instead of "Garmin Device Interface API". You can try this: https://github.com/Tigge/Garmin-Forerunner-610-Extractor.

## Installing

### Easy Install (stable)

Make sure your system has python, pip, and libusb-1.0:

Debian/Ubuntu:
    sudo apt-get install python-pip libusb-1.0-0
Fedora:
    yum install python-pip libusb

You may also want to install python-lxml. If you skip this step pip will need to build from source (which requires libxml2 libxslt1 dev packages):

Debian/Ubuntu:
    sudo apt-get install python-lxml
Fedora:
    sudo yum install python-lxml

Once prerequisites are installed you can install python-ant-downloader from PyPi:

    sudo pip install python-ant-downloader

### Manual Install (stable/unstable)

You can either clone the git project:

    git clone git://github.com/braiden/python-ant-downloader.git

Or download a stable build from [PyPi](http://pypi.python.org/pypi/python_ant_downloader) or [Github tags](https://github.com/braiden/python-ant-downloader/tags)

##### Prerequisites

 * Python 2.6+
 * [pyusb 1.0](https://github.com/walac/pyusb) - latest version from github is recommended. 0.4 will not work.
 * [poster](http://pypi.python.org/pypi/poster) - only if you enable upload to garmin connect
 * [argparse](http://pypi.python.org/pypi/argparse) - if < python 2.7
 * [lxml](http://pypi.python.org/pypi/lxml)
 * [setuptools](http://pypi.python.org/pypi/setuptools)
 * pyserial (required for older hardware revisions of USB ANT Stick)

On Ubuntu most of these dependencies can be satisfied with:

    apt-get install python python-lxml python-pkg-resources python-poster python-serial

But, you will still need to download pyusb from github or PyPi.

Fedora:
    sudo yum install python python-lxml pyusb
    sudo easy_install poster


## Running

	$ ant-downloader --help
	
	usage: ant-downloader [-h] [--config f] [--daemon] [--verbose]
	optional arguments:
	  -h, --help        show this help message and exit
	  --config f, -c f  use provided configuration, defaults ~/.antd/antd.cfg,
	  --daemon, -d      run in continuous search mode downloading data from any
	                    availible devices, WILL NOT PAIR WITH NEW DEVICES
	  --verbose, -v     enable all debugging output, NOISY: see config file to
	                    selectively enable loggers

### First Time

Make sure you have permission to access the USB device. Add a text file with one of the following to /etc/udev/rules.d/99-garmin.rules.

On Ubuntu 10.04 (or other other older distros):

	SUBSYSTEM=="usb", SYSFS{idVendor}=="0fcf", SYSFS{idProduct}=="1008", MODE="666"

On Ubuntu 12.04, Fedora 19 (or other distros running newer udev):

	SUBSYSTEM=="usb", ATTR{idVendor}=="0fcf", ATTR{idProduct}=="1008", MODE="666"

The first time you run the program it will need to pair with your GPS device. Make sure the the GPS unit is awake (press a button), and make sure pairing is enabled. Then just run <code>ant-downloader</code>. When prompted accept the pairing request on your GPS device. Once request is accepted a key is saved and you should not need to pair again.

You may also choose to enable "Force Downloads" on your device. This will cause all old data to be downloaded. WARNING, It will also upload all data to Garmin Connect.

Also the device must not be claimed by the usbserial kernel module. 
if you get an error and dmesg says

	 usb 3-1.2: usbfs: interface 0 claimed by usbserial_generic while 'ant-downloader' sets config #1

try unloading the 'usbserial' kernel module. 

### Configuration

See antd.cfg from configuration options including where files are saved, and Garmin Connect login details. The file will be created in ~/.antd the first time you run the program.



CHANGES
- brodykenrick
--- Modifications to add various Data Pages including RequestPage, HeartRatePages ---
--- Modifications to add much more human readable logging that does lookups on message ids, codes etc and dissects 'data' payloads and applies a page unpacker on them ---

e.g.
[Thread-2]	2013-08-09 16:55:25,410	DEBUG	Processing reply. RECV_BROADCAST_DATA(channel_number=0, data=0x4301000301000200 -> Beacon{'auth_type': 3, 'pairing_enabled': 0, 'data_available': 0, 'device_state': 0, 'period': 1, 'descriptor': 131073, 'upload_enabled': 0, 'data_page_id': 67, 'data': '', 'status_2': 0, 'status_1': 1})
[Thread-1]	2013-08-09 16:55:09,444	DEBUG	Processing reply. CHANNEL_EVENT(channel_number=0, msg_id=MESG_NETWORK_KEY_ID[70], msg_code=RESPONSE_NO_ERROR[0])


 - master
    - Add Creator section in tcx. by Carlroth, issue #22
    - cleanup setup.py, shouldn't fail when downloaded from PyPi (once released)
    - add dependency on distribution (to ensure pkg_resource is availible)
    - add dbus notifications (written by ivankelly)
    - added suppport for older USB ANT stick (nRF241AP1 with USB<->Serial)
    - fix for failed garmin update when login name != user name
    - plugin for email upload to Strava
 - 2012-02-25
    - setup tools, automated installer
	- check version# of config file, and generate warning if
	  file might not be backward compatible
	- config file must exist in ~/.antd
 - 2012-02-19
 	- fix bug where burst transfers may fail do to incorrect sequence
	  numbers. (0 should only be used for first packet)
	- merged branch 'multidevice': multi-device suppport
		- old known devices dbm is incompatable, you will need to repair
		- some configuration updated, tcx, raw, etc now write to path
		  which includes device serial number. It is recommended you
		  use latest configuration file. Update if you've copied it 
		  to ~/.antd.
		- pairing is only accepted when not running in daemon mode.
		  This seems to make sense because otherwise, the background
		  process would potentially drain battery of unparied devices
		  which are found in range.
		- updated antfs search code, can search for a device with specfic
		  serial numbers. prerequisites to implementing uploads.


test/test_request_data_page.py
#!/usr/bin/python

#Tests sending a request data page to a device

import sys
import logging
import time

import antd.ant as ant
import antd.hw  as hw

from antd.ant   import AntTimeoutError

from antd.ant   import RequestDataPage, DATA_PAGE_COMMON_MANUFACTURERS_INFO, DATA_PAGE_REQUEST_DATA
from antd.ant   import HeartRatePage0, HeartRatePage1, HeartRatePage2, HeartRatePage3, HeartRatePage4
from antd.antfs import Beacon

logging.basicConfig(
        level=logging.DEBUG,
        out=sys.stderr,
        format="[%(threadName)s]\t%(asctime)s\t%(levelname)s\t%(message)s")
_LOG = logging.getLogger()

#Tested with Garmin Forerunner 410 (acks request but does not send)
#Tested with Garmin Soft HRM (ignores request)
def listen_for_broadcast_then_request_data_page(session, isGPSNotHRM):
        network = session.networks[0]
        if isGPSNotHRM:
            network.set_key("\xa8\xa4\x23\xb9\xf5\x5e\x63\xc1")
        else:
            network.set_key('\xB9\xA5\x21\xFB\xBD\x72\xC3\x45')
        channel = session.channels[0]
        channel.assign(channel_type=0x00, network_number=0)
        if isGPSNotHRM:
            channel.set_id(device_number=0, device_type_id=0, trans_type=0)
            channel.set_period(0x4000)
            channel.set_rf_freq(50)
        else:
            channel.set_id(device_number=0, device_type_id=120, trans_type=0)
            channel.set_period(8070)
            channel.set_rf_freq(57)
            
        channel.set_search_waveform(0x0053)
        channel.set_search_timeout(20)
        channel.open()
        
        _LOG.info("Wait on a broadcast. See if there is something there and set period to matching if we can....")
        broadcast_data = channel.recv_broadcast( timeout=5 )

        beacon = Beacon.unpack(broadcast_data)
        if beacon:
            print "ANT-FS | Beacon : " + str(beacon)
            #Matching the period. Gives cleaner logs.
            period_hz = 2 ** (beacon.period - 1)
            channel_period = 0x8000 / period_hz
            channel.set_period( channel_period )
            
        heart_rate = HeartRatePage0.unpack(broadcast_data)
        if not(heart_rate):
            heart_rate = HeartRatePage1.unpack(broadcast_data)
        if not(heart_rate):
            heart_rate = HeartRatePage2.unpack(broadcast_data)
        if not(heart_rate):
            heart_rate = HeartRatePage3.unpack(broadcast_data)
        if not(heart_rate):
            heart_rate = HeartRatePage4.unpack(broadcast_data)
        if heart_rate:
            if(heart_rate):
                heart_rate_comp = str(heart_rate.computed_heart_rate)
            else:
                heart_rate_comp = "INVALID"
            print "HRM | Heart Rate of " + heart_rate_comp + " BPM : " + str( heart_rate )

        time.sleep(.750)

                
        print "------- " + "Send Common Data Page Request for Page "+str(DATA_PAGE_COMMON_MANUFACTURERS_INFO)+" and wait for Ack" + " -----------"
        from antd.ant import AntTxFailedError
        try:
            req_data_page = RequestDataPage( data_page_number=DATA_PAGE_REQUEST_DATA, reserved1=0xFF, reserved2=0xFF, descriptor_byte1=0x00, descriptor_byte2=0x00, requested_transmission_response=0x80, requested_page_number=DATA_PAGE_COMMON_MANUFACTURERS_INFO, command_id=0x01 )
            channel.send_acknowledged( req_data_page.pack() )
        except AntTxFailedError, e:
            print "---------------- " + "Failure to get an Ack" + " -----------------------"
        else:
            print "---------------- " + "Received Ack" + " -----------------------"

        #I never received a data page response. I was just watching the log here.

        time.sleep(5)
        print "---------------- " + "End of wait" + " -----------------------"
########################



def main():

    dev = hw.UsbHardware()
    core = ant.Core(dev)

    print ">>>>>>>>>>>>>>>>  Garmin (Soft) HRM >>>>>>>>>>>>>>>>>>>>>>>>>>>>>"

    session = ant.Session(core)
    try:
        listen_for_broadcast_then_request_data_page(session, isGPSNotHRM=False)
    except AntTimeoutError:
        _LOG.warning("Timed out - Device not present?.")
    finally:
        try: session.close()
        except: _LOG.warning("Caught exception while resetting system.", exc_info=True)

    print ">>>>>>>>>>>>>>>>>  Garmin Forerunner 410 >>>>>>>>>>>>>>>>>>>>>>>"

    session = ant.Session(core)
    try:
        listen_for_broadcast_then_request_data_page(session, isGPSNotHRM=True)
    except AntTimeoutError:
        _LOG.warning("Timed out - Device not present?.")
    finally:
        try: session.close()
        except: _LOG.warning("Caught exception while resetting system.", exc_info=True)

    print "---------------- " + "Finishing" + " -----------------------"
########################


if __name__ == "__main__":
    print "Script = " + sys.argv[0]
    main()


antd/antfs.py
# Copyright (c) 2012, Braiden Kindt.
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 
#   1. Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
# 
#   2. Redistributions in binary form must reproduce the above
#      copyright notice, this list of conditions and the following
#      disclaimer in the documentation and/or other materials
#      provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER AND CONTRIBUTORS
# ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
# WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.


import random
import struct
import collections
import logging
import time
import os
import socket
import binascii
import ConfigParser

import antd.ant as ant

_log = logging.getLogger("antd.antfs")

ANTFS_HOST_ID = os.getpid() & 0xFFFFFFFF
ANTFS_HOST_NAME = socket.gethostname()[:8]


class Beacon(object):

    DATA_PAGE_ID = 0x43
    STATE_LINK, STATE_AUTH, STATE_TRANSPORT, STATE_BUSY = range(0,4)

    __struct = struct.Struct("<BBBBI")

    @classmethod
    def unpack(cls, msg):
        if msg and ord(msg[0]) == Beacon.DATA_PAGE_ID:
            result = cls()
            result.data_page_id, result.status_1, result.status_2, result.auth_type, result.descriptor = cls.__struct.unpack(msg[:8])
            result.period = 0x07 & result.status_1
            result.pairing_enabled = 0x80 & result.status_1
            result.upload_enabled  = 0x10 & result.status_1
            result.data_available  = 0x20 & result.status_1
            result.device_state    = 0x0f & result.status_2
            result.data = msg[8:]
            return result

    def __str__(self):
        #TODO : Processing of data
        return self.__class__.__name__ + str(self.__dict__)

#Strangely, even though the command data is a sub part of a beacon, the beacon is made the field of the command if it is recognised....
class Command(object):

    DATA_PAGE_ID = 0x44
    LINK, DISCONNECT, AUTH, PING, DIRECT = 0x02, 0x03, 0x04, 0x05, 0x0D

    __struct = struct.Struct("<BB6x")

    @classmethod
    def unpack(cls, msg):
        beacon = Beacon.unpack(msg) 
        if beacon and beacon.data and ord(beacon.data[0]) == Command.DATA_PAGE_ID:
            result = cls()
            result.beacon = beacon
            result.data_page_id, result.command_id = cls.__struct.unpack(beacon.data[:8])
            return result
        else:
            #Command standing on its own (not as sub part of a beacon)
            if ord(msg[0]) == Command.DATA_PAGE_ID:
                result = cls()
                result.data_page_id, result.command_id = cls.__struct.unpack(msg[:8])
                result.data = msg[2:]
                return result

    def __str__(self):
        #TODO : Processing of data
        return self.__class__.__name__ + str(self.__dict__)


class Disconnect(Command):
    
    COMMAND_ID = Command.DISCONNECT

    __struct = struct.Struct("<BB6x")

    def pack(self):
        return self.__struct.pack(self.DATA_PAGE_ID, self.COMMAND_ID)


class Ping(Command):
    
    COMMAND_ID = Command.PING

    __struct = struct.Struct("<BB6x")

    def pack(self):
        return self.__struct.pack(self.DATA_PAGE_ID, self.COMMAND_ID)


class Link(Command):

    COMMAND_ID = Command.LINK 

    __struct = struct.Struct("<BBBBI")

    def __init__(self, freq, period, host_id=ANTFS_HOST_ID):
        self.frequency = freq
        self.period = period
        self.host_id = host_id

    def pack(self):
        return self.__struct.pack(self.DATA_PAGE_ID, self.COMMAND_ID, self.frequency, self.period, self.host_id)
    

class Auth(Command):

    COMMAND_ID = Command.AUTH

    OP_PASS_THRU, OP_CLIENT_SN, OP_PAIR, OP_PASSKEY = range(0, 4)
    RESPONSE_NA, RESPONSE_ACCEPT, RESPONSE_REJECT = range(0, 3)

    __struct = struct.Struct("<BBBBI")

    def __init__(self, op_id=None, auth_string="", host_id=ANTFS_HOST_ID):
        self.op_id = op_id
        self.auth_string = auth_string
        self.host_id = host_id

    def pack(self):
        return self.__struct.pack(self.DATA_PAGE_ID, self.COMMAND_ID, self.op_id, len(self.auth_string), self.host_id) + self.auth_string
    
    @classmethod
    def unpack(cls, msg):
        auth = super(Auth, cls).unpack(msg)
        if auth and auth.command_id & 0x7F == Auth.COMMAND_ID:
            try:
                #This only works for Auth as a sub part of a beacon....
                data_page_id, command_id, auth.response_type, auth_string_length, auth.client_id = cls.__struct.unpack(auth.beacon.data[:8])
                auth.auth_string = auth.beacon.data[8:8 + auth_string_length]
            except:
                #If we want to unpack Auth just as data, that needs to use msg
                data_page_id, command_id, auth.response_type, auth_string_length, auth.client_id = cls.__struct.unpack(msg[:8])
                auth.auth_string = msg[8:8 + auth_string_length]
            return auth


class GarminSendDirect(Command):
    
    COMMAND_ID = Command.DIRECT

    __struct = struct.Struct("<BBHHH")

    def __init__(self, data="", fd=0xFFFF, offset=0x0000):
        self.fd = fd
        self.offset = offset
        self.data = data
        self.blocks = (len(data) - 1) // 8

    def pack(self):
        return self.__struct.pack(self.DATA_PAGE_ID, self.COMMAND_ID, self.fd, self.offset, self.blocks) + self.data
    
    @classmethod
    def unpack(cls, msg):
        direct = super(GarminSendDirect, cls).unpack(msg)
        if direct and direct.command_id & 0x7F == GarminSendDirect.COMMAND_ID:
            data_page_id, command_id, direct.fd, direct.offset, direct.blocks = cls.__struct.unpack(direct.beacon.data[:8])
            direct.data = direct.beacon.data[8:8 + 8 * direct.blocks]
            return direct


class KnownDeviceDb(object):

    def __init__(self, file = None):
        self.file = file
        self.key_by_device_id = dict()
        self.device_id_by_ant_device_number = dict()
        self.cfg = ConfigParser.SafeConfigParser()
        if file: self.cfg.read([file])
        for section in self.cfg.sections():
            device_id = int(section, 0)
            try: self.key_by_device_id[device_id] = binascii.unhexlify(self.cfg.get(section, "key"))
            except ConfigParser.NoOptionError: pass
            try: self.device_id_by_ant_device_number[int(self.cfg.get(section, "device_number"), 0)] = device_id
            except ConfigParser.NoOptionError: pass

    def get_key(self, device_id):
        return self.key_by_device_id.get(device_id, None)

    def add_key(self, device_id, key):
        self.add_to_cfg(device_id, "key", key.encode("hex"))

    def get_device_id(self, ant_device_number):
        return self.device_id_by_ant_device_number.get(ant_device_number, None)

    def add_device_id(self, ant_device_number, device_id):
        self.add_to_cfg(device_id, "device_number", hex(ant_device_number))

    def delete_device(self, device_id):
        section = "0x%08x" % device_id 
        try: self.cfg.remove_section(section)
        except ConfigParser.NoSectionError: pass
        else:
            if self.file:
                with open(self.file, "w") as file:
                    self.cfg.write(file)
        
    def add_to_cfg(self, device_id, key, value):
        section = "0x%08x" % device_id 
        try: self.cfg.add_section(section)
        except ConfigParser.DuplicateSectionError: pass
        self.cfg.set(section, key, value)
        if self.file:
            with open(self.file, "w") as file:
                self.cfg.write(file)


class Host(object):

    search_network_key = "\xa8\xa4\x23\xb9\xf5\x5e\x63\xc1"
    search_freq = 50
    search_period = 0x1000
    search_timeout = 255
    search_waveform = 0x0053

    transport_freqs = [3, 7, 15, 20, 25, 29, 34, 40, 45, 49, 54, 60, 65, 70, 75, 80]
    transport_period = 0b100
    transport_timeout = 2

    def __init__(self, ant_session, known_client_keys=None):
        self.ant_session = ant_session
        self.known_client_keys = known_client_keys if known_client_keys is not None else KnownDeviceDb()

    def close(self):
        self.channel.send_acknowledged(Disconnect().pack(), direct=True)
        self.ant_session.close()

    def disconnect(self):
        try:
            beacon = Beacon.unpack(self.channel.recv_broadcast(.5))
        except ant.AntTimeoutError:
            pass
        else:
            self.channel.send_acknowledged(Disconnect().pack(), direct=True)
            self.channel.close()

    def ping(self):
        self.channel.write(Ping().pack())

    def search(self, search_timeout=60, device_id=None, include_unpaired_devices=False, include_devices_with_no_data=False):
        """
        Search for devices. If device_id is None return the first device
        which has data available. Unless include_unpaired_devices = True
        only devices for which we know secret key will be returned.

        If device_id is non-None, we will continue to search for device who's
        ANT device_id matchers the requested value. If found the device
        is returned regardless of whether it has data or not.
        include_unpaired_devices is ignored when device_id is provided.
        """
        timeout = time.time() + search_timeout
        while time.time() < timeout:
            try:
                # if we didn't find a device, maybe another is in range?
                # restart search every time. Once a device is tracking
                # we don't get any more hits. So, just keep re-openning
                # channel until we find device we're looking for.
                # TODO could implement AP2 filters, but this logic maintains
                # support for older devices.
                self._open_antfs_search_channel()
                # wait to recv beacon from device
                beacon = Beacon.unpack(self.channel.recv_broadcast(timeout=timeout - time.time()))
            except ant.AntTimeoutError:
                # ignore timeout error
                pass
            else:
                tracking_device_number = self.channel.get_id().device_number
                tracking_device_id = self.known_client_keys.get_device_id(tracking_device_number)
                # check if event was a beacon
                if beacon:
                    _log.debug("Got ANT-FS Beacon. device_number=0x%04x %s", tracking_device_number, beacon)
                    # and if device is a state which will accept our link
                    if  beacon.device_state != Beacon.STATE_LINK:
                        _log.warning("Device busy, not ready for link. device_number=0x%04x state=%d.",
                                tracking_device_number, beacon.device_state)
                    # are we looking for a sepcific device
                    if device_id is not None:
                        if device_id == tracking_device_id:
                            # the device exactly matches the one we're looking for
                            self.beacon = beacon
                            self.device_id = tracking_device_id
                            return beacon
                        else:
                            # a specific device id was request, but is not the one
                            # currently linked, try again. FIXME should really implement
                            # AP2 filters
                            _log.debug("Found device, but device_id does not match. 0x%08x != 0x%08x", tracking_device_id or 0, device_id)
                            continue
                    elif not include_unpaired_devices and tracking_device_id is None:
                        # requested not to return unpared devices
                        # but the one linked is unkown.
                        # FIXME add device to AP2 filter and contiue search
                        _log.debug("Found device, but paring not enabled. device_number=0x%04x", tracking_device_number)
                        continue
                    elif not beacon.data_available and not include_devices_with_no_data:
                        _log.debug("Found device, but no new data for download. device_number=0x%04x", tracking_device_number)
                    else:
                        self.beacon = beacon
                        self.device_id = tracking_device_id # may be None
                        return beacon
        
    def link(self):
        """
        Attempt to create an ANTFS link with the device
        who's beacon was most recently returned by search().
        If this channel is not tracking, the operation will
        block until a device is found, and attempt a link.
        Operation will raise a timeout exception if device
        does not reply in time our if an attempt was made
        to link while channel was not tracking.
        """
        # make sure our message period matches the target
        _log.debug("Setting period to match device, hz=%d",  2 ** (self.beacon.period - 1))
        self._configure_antfs_period(self.beacon.period)
        # wait for channel to sync
        Beacon.unpack(self.channel.recv_broadcast(0))
        # send the link command
        link = Link(freq=random.choice(self.transport_freqs), period=self.transport_period)
        _log.debug("Linking with device. freq=24%02dmhz", link.frequency)
        self.channel.send_acknowledged(link.pack())
        # change this channels frequency to match link
        self._configure_antfs_transport_channel(link)
        # block indefinately for the antfs beacon on new freq.
        # (don't need a timeout since channel will auto close if device lost)
        self.beacon = Beacon.unpack(self.channel.recv_broadcast(0))
        # device should be broadcasting our id and ready to accept auth
        assert self.beacon.device_state == Beacon.STATE_AUTH
        return self.beacon

    def auth(self, pair=True, timeout=60):
        """
        Attempt to create an authenticated transport
        with the device we are currently linked. Not
        valid unless device is in link status.
        If a client key is known, transport will be
        opened without user interaction. If key is unknown
        we will attempt to pair with device (which must
        be acknowledged by a human on GPS device.)
        If pairing is not enabled Auth is impossible.
        Error raised if auth is not successful.
        Timeout only applies user interaction during pairing process.
        """
        # get the S/N of client device
        auth_cmd = Auth(Auth.OP_CLIENT_SN)
        self.channel.write(auth_cmd.pack())
        while True:
            auth_reply = Auth.unpack(self.channel.read())
            if auth_reply: break
        _log.debug("Got client auth string. %s", auth_reply)
        # check if the auth key for this device is known
        client_id = auth_reply.client_id
        # property may not have been set yet if new device
        self.device_id = client_id  
        # look up key
        key = self.known_client_keys.get_key(client_id)
        if key:
            _log.debug("Device secret known.")
            auth_cmd = Auth(Auth.OP_PASSKEY, key)
            self.channel.write(auth_cmd.pack())
            while True:
                auth_reply = Auth.unpack(self.channel.read())
                if auth_reply: break
            if auth_reply.response_type == Auth.RESPONSE_ACCEPT:
                _log.debug("Device accepted key.")
            else:
                _log.warning("Device pairing failed. Removing key from db. Try re-pairing.")
                self.known_client_keys.delete_device(client_id)
        elif pair:
            _log.debug("Device unknown, requesting pairing.")
            auth_cmd = Auth(Auth.OP_PAIR, ANTFS_HOST_NAME)
            self.channel.write(auth_cmd.pack())
            while True:
                auth_reply = Auth.unpack(self.channel.read(timeout))
                if auth_reply: break
            if auth_reply.response_type == Auth.RESPONSE_ACCEPT:
                _log.debug("Device paired. key=%s", auth_reply.auth_string.encode("hex"))
                self.known_client_keys.add_key(client_id, auth_reply.auth_string)
                device_number = self.channel.get_id().device_number
                self.known_client_keys.add_device_id(device_number, client_id)
            else:
                _log.warning("Device pairing failed. Request rejected?")
        else:
            _log.warning("Device 0x08%x has data but pairing is disabled and key is unkown.", client_id)
        #confirm the ANT-FS channel is open
        self.beacon = Beacon.unpack(self.channel.recv_broadcast(0))
        assert self.beacon.device_state == Beacon.STATE_TRANSPORT
        return self.beacon

    def write(self, msg):
        direct_cmd = GarminSendDirect(msg)
        self.channel.write(direct_cmd.pack())

    def read(self):
        direct_reply = GarminSendDirect.unpack(self.channel.read())
        return direct_reply.data if direct_reply else None

    def _open_antfs_search_channel(self):
        self.ant_session.reset_system()
        self.channel = self.ant_session.channels[0]
        self.network = self.ant_session.networks[0]
        self._configure_antfs_search_channel()
        self.channel.open()

    def _configure_antfs_search_channel(self):
        self.network.set_key(self.search_network_key)
        self.channel.assign(channel_type=0x00, network_number=self.network.network_number)
        self.channel.set_id(device_number=0, device_type_id=0, trans_type=0)
        self.channel.set_period(self.search_period)
        self.channel.set_search_timeout(self.search_timeout)
        self.channel.set_rf_freq(self.search_freq)
        self.channel.set_search_waveform(self.search_waveform)

    def _configure_antfs_transport_channel(self, link):
        self.channel.set_rf_freq(link.frequency)
        self.channel.set_search_timeout(self.transport_timeout)
        self._configure_antfs_period(link.period)

    def _configure_antfs_period(self, period):
        period_hz = 2 ** (period - 1)
        channel_period = 0x8000 / period_hz
        self.channel.set_period(channel_period)
        

# vim: ts=4 sts=4 et
